File: Streamy/Streamy/Helper/StreamManager.m
//
//  StreamManager.m
//  Streamy
//
//  Created by Arthur Myronenko on 3/3/16.
//  Copyright © 2016 Arthur Myronenko. All rights reserved.
//

#import "StreamManager.h"
#import <AVFoundation/AVFoundation.h>

@interface StreamManager () <AVCaptureVideoDataOutputSampleBufferDelegate>

@property (strong, nonatomic) AVCaptureSession *session;
@property (strong, nonatomic) AVCaptureVideoDataOutput *dataOutput;

@property (weak, nonatomic) id<StreamManagerDelegate> delegate;

@end

@implementation StreamManager

- (void)startCaptureWithDelegate:(id<StreamManagerDelegate>)delegate {
    self.delegate = delegate;
    
    self.session = [[AVCaptureSession alloc] init];
    self.session.sessionPreset = AVCaptureSessionPreset960x540;
    
    CGDirectDisplayID displayId = kCGDirectMainDisplay;
    
    AVCaptureScreenInput *input = [[AVCaptureScreenInput alloc] initWithDisplayID:displayId];
    if (!input) {
        self.session = nil;
        return;
    }
    if ([self.session canAddInput:input]) {
        [self.session addInput:input];
    }
    
    self.dataOutput = [[AVCaptureVideoDataOutput alloc] init];
    if ([self.session canAddOutput:self.dataOutput]) {
        [self.session addOutput:self.dataOutput];
    }
    
    [self.session startRunning];
    
    [self.dataOutput setSampleBufferDelegate:self queue:dispatch_get_main_queue()];
}

- (void)stop {
    [self.session stopRunning];
}

#pragma mark - AVCaptureVideoDataOutputSampleBufferDelegate

- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection {
    CVPixelBufferRef pixelBuffer = (CVPixelBufferRef)CMSampleBufferGetImageBuffer(sampleBuffer);
    CIImage *ciImage = [CIImage imageWithCVPixelBuffer:pixelBuffer];
    CIContext *context = [CIContext contextWithOptions:nil];
    CGFloat width = CVPixelBufferGetWidth(pixelBuffer);
    CGFloat height = CVPixelBufferGetHeight(pixelBuffer);
    
    CGImageRef myImage = [context
                          createCGImage:ciImage
                          fromRect:CGRectMake(0, 0, width, height)];
    
    NSImage *result = [[NSImage alloc] initWithCGImage:myImage size: CGSizeMake(width, height)];
    CGImageRelease(myImage);

    [self.delegate streamManager:self didCaptureFrame:result];
}

@end

File: Streamy/Streamy/StreamManager.swift
//
//  StreamManager.swift
//  Streamy
//
//  Created by Arthur Myronenko on 4/17/16.
//  Copyright © 2016 Arthur Myronenko. All rights reserved.
//

import AVFoundation

protocol StreamManagerDelegate {
    func streamManager(streamManager: StreamManager, didCaptureFrame: NSImage)
}

class StreamManager: NSObject {
    private let session: AVCaptureSession
    private let dataOutput: AVCaptureVideoDataOutput
    private var delegate: StreamManagerDelegate?
    
    override init() {
        session = AVCaptureSession()
        session.sessionPreset = AVCaptureSessionPreset960x540
        
        let displayId = CGMainDisplayID()
        
        guard let input = AVCaptureScreenInput(displayID: displayId) else {
            fatalError()
        }
        
        if (session.canAddInput(input)) {
            session.addInput(input)
        }
        
        dataOutput = AVCaptureVideoDataOutput()
        if (session.canAddOutput(dataOutput)) {
            session.addOutput(dataOutput)
        }
        
        super.init()
    }
    
    func startCaptureWithDelegate(delegate: StreamManagerDelegate?) {
        self.delegate = delegate
        session.startRunning()
        dataOutput.setSampleBufferDelegate(self, queue: dispatch_get_main_queue())
    }
    
    func stop() {
        session.stopRunning()
    }
}

extension StreamManager: AVCaptureVideoDataOutputSampleBufferDelegate {
    func captureOutput(captureOutput: AVCaptureOutput!,
                       didOutputSampleBuffer sampleBuffer: CMSampleBuffer!,
                       fromConnection connection: AVCaptureConnection!) {
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return print("Error") }
        let ciImage = CIImage(CVImageBuffer: pixelBuffer)
        let context = CIContext(options: nil)
        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)
        
        let frame = context.createCGImage(ciImage, fromRect: CGRect(x: 0, y: 0, width: width, height: height))
        let result = NSImage(CGImage: frame, size: CGSize(width: width, height: height))
        
        self.delegate?.streamManager(self, didCaptureFrame: result)
    }
}
